File: config/config-java/src/test/java/quan/config/quest/QuestConfig.java
package quan.config.quest;

import java.util.*;
import com.alibaba.fastjson.*;
import quan.config.*;
import quan.config.item.Reward;

/**
* 任务<br/>
* Created by 自动生成
*/
@SuppressWarnings({"unchecked"})
public class QuestConfig extends Config {

    //ID
    protected final int id;

    //名字
    protected final String name;

    //类型
    protected final QuestType type;

    //任务目标
    protected final int target;

    //奖励
    protected final Reward reward;

    //A1
    protected final int a1;

    //A2
    protected final int a2;

    //B1
    protected final int b1;

    //B2
    protected final boolean b2;

    //C1
    protected final String c1;

    //C2
    protected final int c2;

    //C3
    protected final int c3;

    //D1
    protected final String d1;

    //D2
    protected final int d2;

    //D3
    protected final int d3;

    //S1
    protected final Set<Integer> s1;

    //L1
    protected final List<Integer> l1;

    //M1
    protected final Map<Integer, Integer> m1;


    public QuestConfig(JSONObject json) {
        super(json);

        id = json.getIntValue("id");
        name = json.getOrDefault("name", "").toString();

        String $type = json.getString("type");
        if ($type != null) {
            type = QuestType.valueOf($type);
        } else {
            type = null;
        }

        target = json.getIntValue("target");

        JSONObject $reward = json.getJSONObject("reward");
        if ($reward != null) {
            reward = new Reward($reward);
        } else {
            reward = null;
        }

        a1 = json.getIntValue("a1");
        a2 = json.getIntValue("a2");
        b1 = json.getIntValue("b1");
        b2 = json.getBooleanValue("b2");
        c1 = json.getOrDefault("c1", "").toString();
        c2 = json.getIntValue("c2");
        c3 = json.getIntValue("c3");
        d1 = json.getOrDefault("d1", "").toString();
        d2 = json.getIntValue("d2");
        d3 = json.getIntValue("d3");

        JSONArray $s1$1 = json.getJSONArray("s1");
        Set<Integer> $s1$2 = new HashSet<>();
        if ($s1$1 != null) {
            for (int i = 0; i < $s1$1.size(); i++) {
                $s1$2.add($s1$1.getInteger(i));
            }
        }
        s1 = Collections.unmodifiableSet($s1$2);

        JSONArray $l1$1 = json.getJSONArray("l1");
        List<Integer> $l1$2 = new ArrayList<>();
        if ($l1$1 != null) {
            for (int i = 0; i < $l1$1.size(); i++) {
                $l1$2.add($l1$1.getInteger(i));
            }
        }
        l1 = Collections.unmodifiableList($l1$2);

        JSONObject $m1$1 = json.getJSONObject("m1");
        Map<Integer, Integer> $m1$2 = new HashMap();
        if ($m1$1 != null) {
            for (String $m1$Key : $m1$1.keySet()) {
                $m1$2.put(Integer.valueOf($m1$Key), $m1$1.getInteger($m1$Key));
            }
        }
        m1 = Collections.unmodifiableMap($m1$2);
    }

    /**
     * ID
     */
    public final int getId() {
        return id;
    }

    /**
     * 名字
     */
    public final String getName() {
        return name;
    }

    /**
     * 类型
     */
    public final QuestType getType() {
        return type;
    }

    /**
     * 任务目标
     */
    public final int getTarget() {
        return target;
    }

    /**
     * 奖励
     */
    public final Reward getReward() {
        return reward;
    }

    /**
     * A1
     */
    public final int getA1() {
        return a1;
    }

    /**
     * A2
     */
    public final int getA2() {
        return a2;
    }

    /**
     * B1
     */
    public final int getB1() {
        return b1;
    }

    /**
     * B2
     */
    public final boolean getB2() {
        return b2;
    }

    /**
     * C1
     */
    public final String getC1() {
        return c1;
    }

    /**
     * C2
     */
    public final int getC2() {
        return c2;
    }

    /**
     * C3
     */
    public final int getC3() {
        return c3;
    }

    /**
     * D1
     */
    public final String getD1() {
        return d1;
    }

    /**
     * D2
     */
    public final int getD2() {
        return d2;
    }

    /**
     * D3
     */
    public final int getD3() {
        return d3;
    }

    /**
     * S1
     */
    public final Set<Integer> getS1() {
        return s1;
    }

    /**
     * L1
     */
    public final List<Integer> getL1() {
        return l1;
    }

    /**
     * M1
     */
    public final Map<Integer, Integer> getM1() {
        return m1;
    }


    @Override
    protected QuestConfig create(JSONObject json) {
        return new QuestConfig(json);
    }

    @Override
    public String toString() {
        return "QuestConfig{" +
                "id=" + id +
                ",name='" + name + '\'' +
                ",type=" + type +
                ",target=" + target +
                ",reward=" + reward +
                ",a1=" + a1 +
                ",a2=" + a2 +
                ",b1=" + b1 +
                ",b2=" + b2 +
                ",c1='" + c1 + '\'' +
                ",c2=" + c2 +
                ",c3=" + c3 +
                ",d1='" + d1 + '\'' +
                ",d2=" + d2 +
                ",d3=" + d3 +
                ",s1=" + s1 +
                ",l1=" + l1 +
                ",m1=" + m1 +
                '}';

    }


    private volatile static List<QuestConfig> configs = new ArrayList<>();

    private volatile static Map<Integer, Map<Integer, QuestConfig>> composite1Configs = new HashMap<>();

    private volatile static Map<Integer, Map<Boolean, List<QuestConfig>>> composite2Configs = new HashMap<>();

    private volatile static Map<String, Map<Integer, Map<Integer, QuestConfig>>> composite3Configs = new HashMap<>();

    private volatile static Map<String, Map<Integer, Map<Integer, List<QuestConfig>>>> composite4Configs = new HashMap<>();

    //ID
    private volatile static Map<Integer, QuestConfig> idConfigs = new HashMap<>();

    //类型
    private volatile static Map<QuestType, List<QuestConfig>> typeConfigs = new HashMap<>();

    public static List<QuestConfig> getConfigs() {
        return configs;
    }

    public static Map<Integer, Map<Integer, QuestConfig>> getComposite1Configs() {
        return composite1Configs;
    }

    public static Map<Integer, QuestConfig> getByComposite1(int a1) {
        return composite1Configs.getOrDefault(a1, Collections.emptyMap());
    }

    public static QuestConfig getByComposite1(int a1, int a2) {
        return getByComposite1(a1).get(a2);
    }

    public static Map<Integer, Map<Boolean, List<QuestConfig>>> getComposite2Configs() {
        return composite2Configs;
    }

    public static Map<Boolean, List<QuestConfig>> getByComposite2(int b1) {
        return composite2Configs.getOrDefault(b1, Collections.emptyMap());
    }

    public static List<QuestConfig> getByComposite2(int b1, boolean b2) {
        return getByComposite2(b1).getOrDefault(b2, Collections.emptyList());
    }

    public static Map<String, Map<Integer, Map<Integer, QuestConfig>>> getComposite3Configs() {
        return composite3Configs;
    }

    public static Map<Integer, Map<Integer, QuestConfig>> getByComposite3(String c1) {
        return composite3Configs.getOrDefault(c1, Collections.emptyMap());
    }

    public static Map<Integer, QuestConfig> getByComposite3(String c1, int c2) {
        return getByComposite3(c1).getOrDefault(c2, Collections.emptyMap());
    }

    public static QuestConfig getByComposite3(String c1, int c2, int c3) {
        return getByComposite3(c1, c2).get(c3);
    }

    public static Map<String, Map<Integer, Map<Integer, List<QuestConfig>>>> getComposite4Configs() {
        return composite4Configs;
    }

    public static Map<Integer, Map<Integer, List<QuestConfig>>> getByComposite4(String d1) {
        return composite4Configs.getOrDefault(d1, Collections.emptyMap());
    }

    public static Map<Integer, List<QuestConfig>> getByComposite4(String d1, int d2) {
        return getByComposite4(d1).getOrDefault(d2, Collections.emptyMap());
    }

    public static List<QuestConfig> getByComposite4(String d1, int d2, int d3) {
        return getByComposite4(d1, d2).getOrDefault(d3, Collections.emptyList());
    }

    public static Map<Integer, QuestConfig> getIdConfigs() {
        return idConfigs;
    }

    public static QuestConfig getById(int id) {
        return idConfigs.get(id);
    }

    public static Map<QuestType, List<QuestConfig>> getTypeConfigs() {
        return typeConfigs;
    }

    public static List<QuestConfig> getByType(QuestType type) {
        return typeConfigs.getOrDefault(type, Collections.emptyList());
    }


    public static List<String> index(List<QuestConfig> configs) {
        Map<Integer, Map<Integer, QuestConfig>> composite1Configs = new HashMap<>();
        Map<Integer, Map<Boolean, List<QuestConfig>>> composite2Configs = new HashMap<>();
        Map<String, Map<Integer, Map<Integer, QuestConfig>>> composite3Configs = new HashMap<>();
        Map<String, Map<Integer, Map<Integer, List<QuestConfig>>>> composite4Configs = new HashMap<>();
        Map<Integer, QuestConfig> idConfigs = new HashMap<>();
        Map<QuestType, List<QuestConfig>> typeConfigs = new HashMap<>();

        List<String> errors = new ArrayList<>();
        QuestConfig oldConfig;

        for (QuestConfig config : configs) {
            oldConfig = composite1Configs.computeIfAbsent(config.a1, k -> new HashMap<>()).put(config.a2, config);
            if (oldConfig != null) {
                String repeatedConfigs = config.getClass().getSimpleName();
                if (oldConfig.getClass() != config.getClass()) {
                    repeatedConfigs += "," + oldConfig.getClass().getSimpleName();
                }
                errors.add(String.format("配置[%s]有重复数据[%s,%s = %s,%s]", repeatedConfigs, "a1", "a2", config.a1, config.a2));
            }

            composite2Configs.computeIfAbsent(config.b1, k -> new HashMap<>()).computeIfAbsent(config.b2, k -> new ArrayList<>()).add(config);

            oldConfig = composite3Configs.computeIfAbsent(config.c1, k -> new HashMap<>()).computeIfAbsent(config.c2, k -> new HashMap<>()).put(config.c3, config);
            if (oldConfig != null) {
                String repeatedConfigs = config.getClass().getSimpleName();
                if (oldConfig.getClass() != config.getClass()) {
                    repeatedConfigs += "," + oldConfig.getClass().getSimpleName();
                }
                errors.add(String.format("配置[%s]有重复数据[%s,%s,%s = %s,%s,%s]", repeatedConfigs, "c1", "c2", "c3", config.c1, config.c2, config.c3));
            }

            composite4Configs.computeIfAbsent(config.d1, k -> new HashMap<>()).computeIfAbsent(config.d2, k -> new HashMap<>()).computeIfAbsent(config.d3, k -> new ArrayList<>()).add(config);

            oldConfig = idConfigs.put(config.id, config);
            if (oldConfig != null) {
                String repeatedConfigs = config.getClass().getSimpleName();
                if (oldConfig.getClass() != config.getClass()) {
                    repeatedConfigs += "," + oldConfig.getClass().getSimpleName();
                }
                errors.add(String.format("配置[%s]有重复数据[%s = %s]", repeatedConfigs, "id", config.id));
            }

            typeConfigs.computeIfAbsent(config.type, k -> new ArrayList<>()).add(config);
        }

        QuestConfig.configs = Collections.unmodifiableList(configs);
        QuestConfig.composite1Configs = unmodifiableMap(composite1Configs);
        QuestConfig.composite2Configs = unmodifiableMap(composite2Configs);
        QuestConfig.composite3Configs = unmodifiableMap(composite3Configs);
        QuestConfig.composite4Configs = unmodifiableMap(composite4Configs);
        QuestConfig.idConfigs = unmodifiableMap(idConfigs);
        QuestConfig.typeConfigs = unmodifiableMap(typeConfigs);

        return errors;
    }

}

File: config/ConfigCS/Test/Quest/QuestConfig.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using Newtonsoft.Json.Linq;
using ConfigCS;
using ConfigCS.Test.Item;

namespace ConfigCS.Test.Quest
{
    /// <summary>
	/// 任务<br/>
	/// Created by 自动生成
	/// </summary>
    public class QuestConfig : Config
    {

        /// <summary>
        /// ID
        /// </summary>
        public readonly int Id;

        /// <summary>
        /// 名字
        /// </summary>
        public readonly string Name;

        /// <summary>
        /// 类型
        /// </summary>
        public readonly QuestType Type;

        /// <summary>
        /// 任务目标
        /// </summary>
        public readonly int Target;

        /// <summary>
        /// 奖励
        /// </summary>
        public readonly Reward Reward;

        /// <summary>
        /// A1
        /// </summary>
        public readonly int A1;

        /// <summary>
        /// A2
        /// </summary>
        public readonly int A2;

        /// <summary>
        /// B1
        /// </summary>
        public readonly int B1;

        /// <summary>
        /// B2
        /// </summary>
        public readonly bool B2;

        /// <summary>
        /// C1
        /// </summary>
        public readonly string C1;

        /// <summary>
        /// C2
        /// </summary>
        public readonly int C2;

        /// <summary>
        /// C3
        /// </summary>
        public readonly int C3;

        /// <summary>
        /// D1
        /// </summary>
        public readonly string D1;

        /// <summary>
        /// D2
        /// </summary>
        public readonly int D2;

        /// <summary>
        /// D3
        /// </summary>
        public readonly int D3;

        /// <summary>
        /// S1
        /// </summary>
        public readonly ISet<int> S1;

        /// <summary>
        /// L1
        /// </summary>
        public readonly IList<int> L1;

        /// <summary>
        /// M1
        /// </summary>
        public readonly IDictionary<int, int> M1;


        public QuestConfig(JObject json): base(json)
        {
            Id = json["id"]?.Value<int>()?? default;
            Name = json["name"]?.Value<string>()?? "";
            Type = (QuestType) (json["type"]?.Value<int>()?? default);
            Target = json["target"]?.Value<int>()?? default;
            Reward = json.ContainsKey("reward") ? new Reward(json["reward"].Value<JObject>()) : null;
            A1 = json["a1"]?.Value<int>()?? default;
            A2 = json["a2"]?.Value<int>()?? default;
            B1 = json["b1"]?.Value<int>()?? default;
            B2 = json["b2"]?.Value<bool>()?? default;
            C1 = json["c1"]?.Value<string>()?? "";
            C2 = json["c2"]?.Value<int>()?? default;
            C3 = json["c3"]?.Value<int>()?? default;
            D1 = json["d1"]?.Value<string>()?? "";
            D2 = json["d2"]?.Value<int>()?? default;
            D3 = json["d3"]?.Value<int>()?? default;

            var s11 = json["s1"]?.Value<JArray>();
            var s12 = ImmutableHashSet<int>.Empty;
            if (s11 != null)
            {
                foreach (var s1Value in s11)
                {
                    s12.Add(s1Value.Value<int>());
                }
            }
            S1 = s12;

            var l11 = json["l1"]?.Value<JArray>();
            var l12 = ImmutableList<int>.Empty;
            if (l11 != null)
            {
                foreach (var l1Value in l11)
                {
                    l12.Add(l1Value.Value<int>());
                }
            }
            L1 = l12;

            var m11 = json["m1"]?.Value<JObject>();
            var m12 = ImmutableDictionary<int, int>.Empty;
            if (m11 != null)
            {
                foreach (var m1Prop in m11.Properties())
                {
                    m12.Add(int.Parse(m1Prop.Name), m1Prop.Value<int>());
                }
            }
            M1 = m12;
        }

        protected override Config Create(JObject json) 
        {
            return new QuestConfig(json);
        }


        public override string ToString()
        {
            return "QuestConfig{" +
                   "Id=" + Id +
                   ",Name='" + Name + '\'' +
                   ",Type=" + Type +
                   ",Target=" + Target +
                   ",Reward=" + Reward +
                   ",A1=" + A1 +
                   ",A2=" + A2 +
                   ",B1=" + B1 +
                   ",B2=" + B2 +
                   ",C1='" + C1 + '\'' +
                   ",C2=" + C2 +
                   ",C3=" + C3 +
                   ",D1='" + D1 + '\'' +
                   ",D2=" + D2 +
                   ",D3=" + D3 +
                   ",S1=" + S1 +
                   ",L1=" + L1 +
                   ",M1=" + M1 +
                   '}';
        }


        private static volatile IList<QuestConfig> _configs = new List<QuestConfig>();

        private static volatile IDictionary<int, IDictionary<int, QuestConfig>> _composite1Configs = new Dictionary<int, IDictionary<int, QuestConfig>>();

        private static volatile IDictionary<int, IDictionary<bool, IList<QuestConfig>>> _composite2Configs = new Dictionary<int, IDictionary<bool, IList<QuestConfig>>>();

        private static volatile IDictionary<string, IDictionary<int, IDictionary<int, QuestConfig>>> _composite3Configs = new Dictionary<string, IDictionary<int, IDictionary<int, QuestConfig>>>();

        private static volatile IDictionary<string, IDictionary<int, IDictionary<int, IList<QuestConfig>>>> _composite4Configs = new Dictionary<string, IDictionary<int, IDictionary<int, IList<QuestConfig>>>>();

        /// <summary>
        /// ID
        /// </summary>
        private static volatile IDictionary<int, QuestConfig> _idConfigs = new Dictionary<int, QuestConfig>();

        /// <summary>
        /// 类型
        /// </summary>
        private static volatile IDictionary<QuestType, IList<QuestConfig>> _typeConfigs = new Dictionary<QuestType, IList<QuestConfig>>();

        public static IList<QuestConfig> GetConfigs() 
        {
            return _configs;
        }

        public static IDictionary<int, IDictionary<int, QuestConfig>> GetComposite1Configs() 
        {
            return _composite1Configs;
        }

        public static IDictionary<int, QuestConfig> GetByComposite1(int a1)
        {
            return _composite1Configs.ContainsKey(a1) ? _composite1Configs[a1] : ImmutableDictionary<int, QuestConfig>.Empty;
        }

        public static QuestConfig GetByComposite1(int a1, int a2)
        {
            var composite1Configs = GetByComposite1(a1);
            return composite1Configs.ContainsKey(a1) ? composite1Configs[a2] : null;
        }

        public static IDictionary<int, IDictionary<bool, IList<QuestConfig>>> GetComposite2Configs()
        {
            return _composite2Configs;
        }

        public static IDictionary<bool, IList<QuestConfig>> GetByComposite2(int b1)
        {
            return _composite2Configs.ContainsKey(b1) ? _composite2Configs[b1] : ImmutableDictionary<bool, IList<QuestConfig>>.Empty;
        }

        public static IList<QuestConfig> GetByComposite2(int b1, bool b2)
        {
            var composite2Configs = GetByComposite2(b1);
            return composite2Configs.ContainsKey(b2) ? composite2Configs[b2] : ImmutableList<QuestConfig>.Empty;
        }

        public static IDictionary<string, IDictionary<int, IDictionary<int, QuestConfig>>> GetComposite3Configs()
        {
            return _composite3Configs;
        }

        public static IDictionary<int, IDictionary<int, QuestConfig>> GetByComposite3(string c1)
        {
            return _composite3Configs.ContainsKey(c1) ? _composite3Configs[c1] : ImmutableDictionary<int, IDictionary<int, QuestConfig>>.Empty;
        }

        public static IDictionary<int, QuestConfig> GetByComposite3(string c1, int c2)
        {
            var composite3Configs = GetByComposite3(c1);
            return composite3Configs.ContainsKey(c2) ? composite3Configs[c2] : ImmutableDictionary<int, QuestConfig>.Empty;
        }

        public static QuestConfig GetByComposite3(string c1, int c2, int c3)
        {
            var composite3Configs = GetByComposite3(c1, c2);
            return composite3Configs.ContainsKey(c3) ? composite3Configs[c3] : null;
        }

        public static IDictionary<string, IDictionary<int, IDictionary<int, IList<QuestConfig>>>> GetComposite4Configs()
        {
            return _composite4Configs;
        }

        public static IDictionary<int, IDictionary<int, IList<QuestConfig>>> GetByComposite4(string d1)
        {
            return _composite4Configs.ContainsKey(d1) ? _composite4Configs[d1] : ImmutableDictionary<int, IDictionary<int, IList<QuestConfig>>>.Empty;
        }

        public static IDictionary<int, IList<QuestConfig>> GetByComposite4(string d1, int d2)
        {
            var composite4Configs = GetByComposite4(d1);
            return composite4Configs.ContainsKey(d2) ? composite4Configs[d2] : ImmutableDictionary<int, IList<QuestConfig>>.Empty;
        }

        public static IList<QuestConfig> GetByComposite4(string d1, int d2, int d3)
        {
            var composite4Configs = GetByComposite4(d1, d2);
            return composite4Configs.ContainsKey(d3) ? composite4Configs[d3] : ImmutableList<QuestConfig>.Empty;
        }

        public static IDictionary<int, QuestConfig> GetIdConfigs() 
        {
            return _idConfigs;
        }

        public static QuestConfig GetById(int id)
        {
            return _idConfigs.ContainsKey(id) ? _idConfigs[id] : null;
        }

        public static IDictionary<QuestType, IList<QuestConfig>> GetTypeConfigs() 
        {
            return _typeConfigs;
        }

        public static IList<QuestConfig> GetByType(QuestType type)
        {
            return _typeConfigs.ContainsKey(type) ? _typeConfigs[type] : ImmutableList<QuestConfig>.Empty;
        }


        public static void Index(List<QuestConfig> configs)
        {
            var composite1Configs = new Dictionary<int, IDictionary<int, QuestConfig>>();
            var composite2Configs = new Dictionary<int, IDictionary<bool, IList<QuestConfig>>>();
            var composite3Configs = new Dictionary<string, IDictionary<int, IDictionary<int, QuestConfig>>>();
            var composite4Configs = new Dictionary<string, IDictionary<int, IDictionary<int, IList<QuestConfig>>>>();
            var idConfigs = new Dictionary<int, QuestConfig>();
            var typeConfigs = new Dictionary<QuestType, IList<QuestConfig>>();

            foreach (var config in configs)
            {
                if (!composite1Configs.ContainsKey(config.A1))
                {
                    composite1Configs.Add(config.A1, new Dictionary<int, QuestConfig>());
                }

                composite1Configs[config.A1][config.A2] = config;


                if (!composite2Configs.ContainsKey(config.B1))
                {
                    composite2Configs.Add(config.B1, new Dictionary<bool, IList<QuestConfig>>());
                }

                if (!composite2Configs[config.B1].ContainsKey(config.B2))
                {
                    composite2Configs[config.B1][config.B2] = new List<QuestConfig>();
                }

                composite2Configs[config.B1][config.B2].Add(config);


                if (!composite3Configs.ContainsKey(config.C1))
                {
                    composite3Configs.Add(config.C1, new Dictionary<int, IDictionary<int, QuestConfig>>());
                }

                if (!composite3Configs[config.C1].ContainsKey(config.C3))
                {
                    composite3Configs[config.C1].Add(config.C2, new Dictionary<int, QuestConfig>());
                }

                composite3Configs[config.C1][config.C2][config.C3] = config;


                if (!composite4Configs.ContainsKey(config.D1))
                {
                    composite4Configs.Add(config.D1, new Dictionary<int, IDictionary<int, IList<QuestConfig>>>());
                }

                if (!composite4Configs[config.D1].ContainsKey(config.D2))
                {
                        composite4Configs[config.D1].Add(config.D2, new Dictionary<int, IList<QuestConfig>>());
                }

                if (!composite4Configs[config.D1][config.D2].ContainsKey(config.D3))
                {
                        composite4Configs[config.D1][config.D2].Add(config.D3, new List<QuestConfig>());
                }

                composite4Configs[config.D1][config.D2][config.D3].Add(config);


                idConfigs[config.Id] = config;


                if (!typeConfigs.ContainsKey(config.Type))
                {
                    typeConfigs.Add(config.Type, new List<QuestConfig>());
                }

                typeConfigs[config.Type].Add(config);

            }

            _configs = configs.ToImmutableList();
            _composite1Configs = composite1Configs.ToImmutableDictionary();
            _composite2Configs = composite2Configs.ToImmutableDictionary();
            _composite3Configs = composite3Configs.ToImmutableDictionary();
            _composite4Configs = composite4Configs.ToImmutableDictionary();
            _idConfigs = idConfigs.ToImmutableDictionary();
            _typeConfigs = typeConfigs.ToImmutableDictionary();

        }
    }
}